pkg/operator/operator_test.go

```diff
@@ -0,0 +1,229 @@
+}
+
+func (f *fixture) newFakeOperator() (*Operator, informers.SharedInformerFactory) {
+	f.kubeClient = fake.NewSimpleClientset(f.kubeObjects...)
+	f.osClient = osclientfake.NewSimpleClientset(f.osObjects...)
+
+	informers := informers.NewSharedInformerFactory(f.kubeClient, 0)
+	recorder := record.NewFakeRecorder(50)
+	oprt := New(
+		namespace,
+		name,
+		imagesJSONFile,
+		"",
+		informers.Apps().V1().Deployments(),
+		f.kubeClient,
+		f.osClient,
+		recorder,
+	)
+
+	oprt.deployListerSynced = alwaysReady
+
+	for _, d := range f.dLister {
+		informers.Apps().V1().Deployments().Informer().GetIndexer().Add(d)
+	}
+	return oprt, informers
+}
+
+func (f *fixture) expectCreateDeploymentAction(d *appsv1.Deployment) {
+	f.kubeActions = append(f.kubeActions, core.NewCreateAction(schema.GroupVersionResource{Resource: "deployments"}, namespace, d))
+}
+
+func (f *fixture) expectGetDeploymentAction(d *appsv1.Deployment) {
+	f.kubeActions = append(f.kubeActions, core.NewGetAction(schema.GroupVersionResource{Resource: "deployments"}, namespace, d.Name))
+}
+
+func (f *fixture) expectGetInfrastructureAction(infra *osconfigv1.Infrastructure) {
+	f.osActions = append(f.osActions, core.NewGetAction(schema.GroupVersionResource{Resource: "infrastructures"}, metav1.NamespaceNone, infra.Name))
+}
+
+func (f *fixture) expectCreateClusterOperatorAction(clusteroperator *osconfigv1.ClusterOperator) {
+	f.osActions = append(f.osActions, core.NewCreateAction(schema.GroupVersionResource{Resource: "clusteroperators"}, metav1.NamespaceNone, clusteroperator))
+}
+
+func (f *fixture) expectGetClusterOperatorAction(clusteroperator *osconfigv1.ClusterOperator) {
+	f.osActions = append(f.osActions, core.NewGetAction(schema.GroupVersionResource{Resource: "clusteroperators"}, metav1.NamespaceNone, clusteroperator.Name))
+}
+
+func (f *fixture) expectUpdateSubresourceClusterOperatorAction(clusteroperator *osconfigv1.ClusterOperator) {
+	f.osActions = append(f.osActions, core.NewUpdateSubresourceAction(schema.GroupVersionResource{Resource: "clusteroperators"}, "status", metav1.NamespaceNone, clusteroperator))
+}
+
+func filterInformerActions(actions []core.Action) []core.Action {
+	ret := []core.Action{}
+	for _, action := range actions {
+		if len(action.GetNamespace()) == 0 &&
+			(action.Matches("list", "deployments") ||
+				action.Matches("watch", "deployments")) {
+			continue
+		}
+		ret = append(ret, action)
+	}
+
+	return ret
+}
+
+func (f *fixture) run(expectError bool) {
+	oprt, informers := f.newFakeOperator()
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informers.Start(stopCh)
+
+	err := oprt.sync(fmt.Sprintf("%s/%s", namespace, name))
+	if !expectError && err != nil {
+		f.t.Errorf("error syncing deployment: %v", err)
+	} else if expectError && err == nil {
+		f.t.Error("expected error syncing deployment, got nil")
+	}
+
+	kubeActions := filterInformerActions(f.kubeClient.Actions())
+	for i, action := range kubeActions {
+		if len(f.kubeActions) < i+1 {
```

Comment:
couldn't we just check once `len(kubeActions) !-= len (f.kubeActions)`
can we call this may be `actualKubeActions` for easier readibility?

Reply:
I prefer to have to checks under the unittest to make debugging easier, but I will rename variables